src/main.rs

```diff
@@ -1,3 +1,146 @@
+    maker::Reaction,
+    mid_market_rate::get_btc_dai_mid_market_rate,
+    network::{self, Nectar, Orderbook},
+    Maker, Spread,
+};
+use std::time::Duration;
+
+async fn init_maker(
+    bitcoin_wallet: bitcoin_wallet::Wallet,
+    ethereum_wallet: ethereum_wallet::Wallet,
+) -> Maker {
+    let initial_btc_balance = bitcoin_wallet.balance().await;
+
+    let initial_dai_balance = ethereum_wallet.dai_balance().await;
+
+    // TODO: from config
+    let btc_max_sell: anyhow::Result<bitcoin::Amount> = unimplemented!();
+    let dai_max_sell: anyhow::Result<dai::Amount> = unimplemented!();
+    let btc_fee_reserve: anyhow::Result<bitcoin::Amount> = unimplemented!();
+    let dai_fee_reserve: anyhow::Result<dai::Amount> = unimplemented!();
+
+    let initial_rate = get_btc_dai_mid_market_rate().await;
+
+    // TODO from config
+    let spread: Spread = unimplemented!();
+
+    match (
+        initial_btc_balance,
+        initial_dai_balance,
+        btc_fee_reserve,
+        dai_fee_reserve,
+        btc_max_sell,
+        dai_max_sell,
+        initial_rate,
+    ) {
+        (
+            Ok(initial_btc_balance),
+            Ok(initial_dai_balance),
+            Ok(btc_fee_reserve),
+            Ok(dai_fee_reserve),
+            Ok(btc_max_sell),
+            Ok(dai_max_sell),
+            Ok(initial_rate),
+        ) => Maker::new(
+            initial_btc_balance,
+            initial_dai_balance.into(),
+            btc_fee_reserve,
+            dai_fee_reserve,
+            btc_max_sell,
+            dai_max_sell,
+            initial_rate,
+            spread,
+        ),
+        // TODO better error handling
+        _ => panic!("Maker initialisation failed!"),
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    // TODO: from config or hardcoded
+    let dai_contract_addr: comit::ethereum::Address = unimplemented!();
+
+    // TODO: Proper wallet initialisation from config
+    let bitcoin_wallet =
+        bitcoin_wallet::Wallet::new(unimplemented!(), unimplemented!(), unimplemented!()).unwrap();
+    let ethereum_wallet =
+        ethereum_wallet::Wallet::new(unimplemented!(), unimplemented!(), unimplemented!()).unwrap();
+
+    let maker = init_maker(bitcoin_wallet, ethereum_wallet).await;
+
+    let orderbook = Orderbook;
+    let nectar = Nectar::new(orderbook);
+
+    let mut swarm: libp2p::Swarm<Nectar> = unimplemented!();
+
+    let initial_sell_order = maker.next_sell_order();
+    let initial_buy_order = maker.next_buy_order();
+
+    match (initial_sell_order, initial_buy_order) {
+        (Ok(sell_order), Ok(buy_order)) => {
+            swarm.orderbook.publish(sell_order.into());
+            swarm.orderbook.publish(buy_order.into());
+        }
+        _ => panic!("Unable to publish initial orders!"),
+    }
+
+    loop {
+        let rate = get_btc_dai_mid_market_rate().await;
+        match rate {
+            Ok(new_rate) => maker.update_rate(unimplemented!()), // maker should record timestamp of this
```

Comment:
`// maker should record timestamp of this` why?

Reply:
Outdated, from previous iteration.